### ISSUE TITLE
Invalid NotNullIfNotNull contract is not reported

### ISSUE THREAD

**Issue by alrz** [dotnet/roslyn#44539]
**Version Used**: https://github.com/dotnet/roslyn/commit/ab680e7d4afb50003b487c28aeb7fcb7c89fd7fe

**Steps to Reproduce**:

```cs
using System.Diagnostics.CodeAnalysis;
#nullable enable
public class C {
    public void Main()
    {
        M1(new()).ToString();
        M2(new()).ToString();
        
        M1(null).ToString();
        M2(null).ToString();
    }
    
    [return: NotNullIfNotNull("p")]
    public object? M1(object? p) {
        if (p is null) return null;
        return Nullable(); // no warnings (one expected)
    }
    
    [return: NotNullIfNotNull("p")]
    public object? M2(object? p) {
        if (p is null) return null;
        return M1(p); // no warnings (correct)
    }
    
    object? Nullable() {return null;}
}
```
([sharplab repro](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgAwAEhAjAHQAimEA5gHYD2smAxlFQMIsAmcAIJMIAG1xRMUANwBYAFABiJgFcxYiMDFxScUdriLCAZnIAmUj1IBvRaQfkzhACykAshExMAFAEp7RzsFR1CPCh8mOAB3fz8qABUWAGUYBG8Gf3kQsId3c0iYuMSUtIyswNzKsPcItQ14pNT0pky/bNy8gvqxRtKWto7HAF9q6oBtQgB2EFIAORYYOfUxAElsBaWVnwAiAAcdvwBdatNSFmAAKzgOGAB+cJ8L69uHvb9batCcUh890mkpB6H2mQJWQ1yoOWGi0OiypAA9AigSxSNEIAgmBkoL8WFE9AAPPY3eD8AI5ByjCmkCbTWabaFrDaLRm7A7HU5mZ4kh75J5XHmkd6fanfbC/f6A4HkKZgjQQsKg2p/dqI5GsNEYrGtHE+DgsBAIEnk0JU0LVbmveYrWFwfy2UE9GRU4ZAA=))

**Comment by RikkiGibson** [comment]
It sounds like when we visit a return, we need to check the flow state of any parameters referenced by NotNullIfNotNull, and if the parameter has not-null flow state, we need to require not-null state on the returned expression.

**Comment by alrz** [comment]
>It sounds like when we visit a return, we need to check the flow state of any parameters referenced by NotNullIfNotNull, and if the parameter has not-null flow state, we need to require not-null state on the returned expression.

That would not cover one hundred percent of cases, consider this:

```cs

[return: NotNullIfNotNull("p")]
object? Method(object? p) => null; // still no warning based on above

Method(new()).ToString(); // NRE but no warning
```
Sounds like we should complain if there is no definitive state on the parameter on exit point, and if there is, then we can decide what is correct in which case the stacked scenario becomes interesting.

```cs
[return: NotNullIfNotNull("p")]
object? Method01(object? p) => Method02(p);  // technically ok, but we probably don't track `p` that far?

[return: NotNullIfNotNull("p")]
object? Method02(object? p) { .. }
```

**Comment by RikkiGibson** [comment]
> we should complain if there is no definitive state on the parameter on exit point

I'm not sure what this means. In our analysis, the state of such a parameter at any point is either maybe-null or not-null. I don't think there's anything we can do about a scenario like this:

```cs
[return: NotNullIfNotNull("p")]
object? Method(object? p) => null; // still no warning based on above

Method(new()).ToString(); // NRE but no warning
```

Our analysis is not able to distinguish the above scenario from one like the following:

```cs
[return: NotNullIfNotNull("p")]
object? Method(object? p)
{
    if (p is null)
        return null;
    return ...;
}

Method(new()).ToString(); // NRE but no warning
```

That is, we don't know if the parameter has a maybe-null state because the argument is sometimes null or sometimes non-null, or if it has a maybe-null state because we explicitly checked for null and it's just impossible for it to contain anything else in the given code path.

We also have a precedent with attributes like `[NotNullWhen(bool)]` to warn only if "we see you're definitely doing something wrong", and not warn if "we don't see you're definitely doing something right" 😉 #44080

Tagging @jcouv @cston to see if you have any ideas for how more NotNullIfNotNull bugs could be caught in flow analysis.

**Comment by alrz** [comment]
>>we should complain if there is no definitive state on the parameter on exit point
>
>I'm not sure what this means. In our analysis, the state of such a parameter at any point is either maybe-null or not-null.

I think what I mean by that is that we branch on `p` state somewhere because the contract itself implies that we're checking `p` for null or not-null at some point. (however this will be an exception in stacked scenarios and where we return the arg itself).

Not quite sure if that's feasible to do but that first example is pretty much "doing something wrong" if not all of them.

**Comment by RikkiGibson** [comment]
Fixed in #47649
